go/protomodule/type_conversions.go

```diff
@@ -0,0 +1,249 @@
+			for i := 0; i < listVal.Len(); i++ {
+				starlarkValue, err := scalarValueToStarlark(listVal.Get(i), fieldDesc)
+				if err != nil {
+					return starlark.None, err
+				}
+				out.Append(starlarkValue)
+			}
+			return out, nil
+		} else if val.Interface() == nil {
+			return newProtoRepeated(fieldDesc), nil
+		}
+		return starlark.None, fmt.Errorf("TypeError: cannot convert %T into list", val.Interface())
+	} else if fieldDesc.IsMap() {
+		if mapVal, ok := val.Interface().(protoreflect.Map); ok {
+			out := newProtoMap(fieldDesc.MapKey(), fieldDesc.MapValue())
+			var rangeErr error
+			mapVal.Range(func(k protoreflect.MapKey, v protoreflect.Value) bool {
+				starlarkKey, err := scalarValueToStarlark(protoreflect.Value(k), fieldDesc.MapKey())
+				if err != nil {
+					rangeErr = err
+					return false
+				}
+
+				starlarkValue, err := scalarValueToStarlark(v, fieldDesc.MapValue())
+				if err != nil {
+					rangeErr = err
+					return false
+				}
+
+				out.SetKey(starlarkKey, starlarkValue)
+				return true
+			})
+			if rangeErr != nil {
+				return starlark.None, rangeErr
+			}
+
+			return out, nil
+		} else if val.Interface() == nil {
+			return newProtoMap(fieldDesc.MapKey(), fieldDesc.MapValue()), nil
+		}
+		return starlark.None, fmt.Errorf("TypeError: cannot convert %T into map", val.Interface())
+	}
+
+	return scalarValueToStarlark(val, fieldDesc)
+}
+
+func scalarValueToStarlark(val protoreflect.Value, fieldDesc protoreflect.FieldDescriptor) (starlark.Value, error) {
+	switch fieldDesc.Kind() {
+	case protoreflect.BoolKind:
+		return starlark.Bool(val.Bool()), nil
+	case protoreflect.Int32Kind:
+		return starlark.MakeInt64(val.Int()), nil
+	case protoreflect.Int64Kind:
+		return starlark.MakeInt64(val.Int()), nil
+	case protoreflect.Uint32Kind:
+		return starlark.MakeUint64(val.Uint()), nil
+	case protoreflect.Uint64Kind:
+		return starlark.MakeUint64(val.Uint()), nil
+	case protoreflect.FloatKind:
+		return starlark.Float(val.Float()), nil
+	case protoreflect.DoubleKind:
+		return starlark.Float(val.Float()), nil
+	case protoreflect.StringKind:
+		return starlark.String(val.String()), nil
+	case protoreflect.BytesKind:
+		// Handle []byte ([]uint8) -> string special case.
+		return starlark.String(val.Bytes()), nil
+	case protoreflect.MessageKind:
+		return nil, fmt.Errorf("MessageKind: Unimplemented")
+	}
+
+	return starlark.None, fmt.Errorf("valueToStarlark: Value unuspported: %s\n", string(fieldDesc.FullName()))
+}
+
+// Verify v can act as fieldDesc
+func scalarTypeCheck(fieldDesc protoreflect.FieldDescriptor, v starlark.Value) error {
+	_, err := scalarValueFromStarlark(fieldDesc, v)
+	return err
+}
+
+func typeError(fieldDesc protoreflect.FieldDescriptor, val starlark.Value, scalar bool) error {
+	expectedType := typeName(fieldDesc)
+
+	// FieldDescriptor has the same typeName for []string and string
+	// and typeError needs to distinguish setting a []string = int versus
+	// appending a value in []string
```

Comment:
Thanks for this helpful comment! OOC, why can't `typeName` handle this differentiation? i.e. return `[]string` for lists and `string` for string.

Reply:
I couldn't figure out a cleaner way to do this, but I'm bothered by it. The problem I was running into is that you have the cases:

``` python
msg.list_int_field = "a" # TypeError: value "a" (type "string") can't be assigned to "[]int"
msg.list_int_field = ["a"] # TypeError: value "a" (type "string") can't be assigned to "int"
```

But from just the field descriptor typeName doesn't know if we're trying to say "you were trying to set a string on a int (value inside a list)" versus "you were trying to set a string on a field that is a list"